collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManager.java

```diff
@@ -0,0 +1,284 @@
+package org.odk.collect.android.formentry.backgroundlocation;
+
+import android.location.Location;
+
+import androidx.annotation.NonNull;
+
+import com.google.android.gms.location.LocationListener;
+
+import org.odk.collect.android.R;
+import org.odk.collect.android.location.client.LocationClient;
+import org.odk.collect.android.logic.AuditConfig;
+import org.odk.collect.android.logic.AuditEvent;
+
+/**
+ * Manages background location for the location audit logging and odk:setlocation action features.
+ * Provides precondition checking and user feedback for both features.
+ *
+ * For location audit logging, manages all the audit logging as well as fetching the location using
+ * Google Play Services.
+ *
+ * {@link org.odk.collect.android.logic.actions.setlocation.CollectSetLocationAction} fetches
+ * location for odk:setlocation actions.
+ *
+ * The implementation uses a state machine concept. Public method represent user or system actions
+ * that clients of this class capture. Based on those actions and various preconditions (Google Play
+ * Services available, location permissions granted, etc), the manager's state changes.
+ */
+public class BackgroundLocationManager implements LocationClient.LocationClientListener, LocationListener {
+    @NonNull
+    private BackgroundLocationState currentState;
+
+    @NonNull
+    private LocationClient locationClient;
+
+    @NonNull
+    private LocationListener locationListener;
+
+    @NonNull
+    private BackgroundLocationHelper helper;
+
+    public BackgroundLocationManager(LocationClient locationClient, BackgroundLocationHelper helper) {
```

Comment:
The way `BackgroundLocationHelper` is being used here really reminds me of the ["Delegate" pattern](https://en.wikipedia.org/wiki/Delegation_pattern). Would it be too nerdy to call it a `BackgroundLocationDelegate` instead?

On that, it could be cool to have the `BackgroundLocationManager` depend on an interface rather than a concrete class. That might help make it clearer where the separation is.

Reply:
It's kind of a wacky delegate because it delegates to different objects/classes, right? It's more of a forward-er? Bag-o-methods?

> That might help make it clearer where the separation is.

I don't know what the separation is. 😭 I see `BackgroundLocationHelper` as a messy "in-progress" kind of thing. I don't really think its interface makes any sense as-is. Again, it's just a bag of stuff I needed. Or do you see a reasonable interface hiding in there? I suppose it could be something like `BackgroundLocationPreconditionChecker` and the last three methods could find another home?

I think ideally there would be a way to get to the form controller from the `BackgroundLocationManager`. It currently doesn't get created until the async task launched by the `FormEntryActivity`'s `onCreate` completes. One option might be to pass it in to the `formFinishedLoading` method? I'm actually kind of liking that... then the top three methods could go somewhere else. Another option would be to make it a true singleton and mutate it rather than throwing it away and recreating it every time a new form is loaded. Don't love that.

Reply:
Yeah I think it would be nice to break into a couple of different interfaces but it might just be good first step for it to be one interface/delegate. Then IMO it's easier to look at it without the implementation and think "how should this actually work"?

Reply:
Sorry, I'm not really clear on this. Do you mean just taking all the methods in `BackgroundLocationHelper` and putting them in an interface? If not, do you want to give me a commit?

Reply:
I think actually you're right looking at it again. I think a nice next step (outside the PR) would be to split it into a few different interfaces (as you're kinda saying).